File: src/armv7_cpp_api/example/hello_world.cpp
// Copyright 2026 TianFeiF
// SPDX-License-Identifier: Apache-2.0
//
// C++ hello-world: move the arm to home pose.
// Usage:
//   ros2 run armv7_cpp_api hello_world_cpp
#include <cstdio>

#include "armv7_cpp_api/client.hpp"

int main()
{
  rclcpp::init(0, nullptr);
  {
    armv7_cpp_api::Armv7Client arm;
    if (!arm.wait_for_joint_state(std::chrono::seconds(5))) {
      std::fprintf(stderr, "no /joint_states received in 5 s\n");
      return 1;
    }
    auto current = arm.get_joint_state();
    std::printf("current joints: ");
    for (auto q : *current) {
      std::printf("%.3f ", q);
    }
    std::printf("\n");

    armv7_cpp_api::JointVector home{};   // zero-initialised = all-zeros
    auto fut = arm.move_to_joint(home, 3.0);
    if (fut->wait(std::chrono::seconds(15))) {
      std::printf("home reached.\n");
    } else {
      std::printf("move failed or timed out.\n");
    }
  }
  rclcpp::shutdown();
  return 0;
}

File: src/armv7_examples/armv7_examples/hello_world.py
# Copyright 2026 TianFeiF
# SPDX-License-Identifier: Apache-2.0
"""hello_world — move the arm to home pose (all zeros), then to a 30 deg pose.

Run:
    ros2 run armv7_examples hello_world
"""
from __future__ import annotations

import sys

import rclpy

from armv7_py import Armv7Client
from armv7_py.client import home_pose, small_demo_pose


def main(argv=None) -> int:
    rclpy.init(args=argv)
    arm = Armv7Client()
    try:
        if not arm.wait_for_joint_state(timeout_sec=5.0):
            print("ERR: no /joint_states within 5 s. "
                  "Is `ros2 launch armv7_bringup arm.launch.py` running?",
                  file=sys.stderr)
            return 1

        print(f"current joints: "
              f"{['%.3f' % q for q in arm.get_joint_state()]}")
        tcp = arm.get_tcp_pose()
        if tcp:
            print(f"current TCP   : "
                  f"({tcp.x:.3f}, {tcp.y:.3f}, {tcp.z:.3f})")

        print("→ moving to home pose...")
        if not arm.move_to_joint(home_pose(), duration_sec=3.0).wait(timeout_sec=15.0):
            print("ERR: move-to-home failed", file=sys.stderr)
            return 2

        print("→ moving to 30° pose...")
        if not arm.move_to_joint(small_demo_pose(), duration_sec=3.0).wait(timeout_sec=15.0):
            print("ERR: move-to-demo failed", file=sys.stderr)
            return 2

        print("→ returning home...")
        arm.move_to_joint(home_pose(), duration_sec=3.0).wait(timeout_sec=15.0)
        print("done.")
        return 0
    finally:
        arm.shutdown()
        rclpy.shutdown()


if __name__ == '__main__':
    sys.exit(main())
